### PR TITLE
[User Service] Admin Management + Authorisation

### PR DESCRIPTION
**Pull request by gongg21**
## Updates in this PR

**FE:**
1. Added `/manage-users` for admin view only and added/updated all relevant components (took reference to `questions` table), including table components, modals, new store, query keys, user DTOs, etc.
2. Minor updates in question components including naming of components and use of stores.
3. Create, update, and deletion buttons for questions can now only be seen by admins (to revert/bypass if it helps with development)
4. Updated `disabled` prop for buttons in modals
5. Update error handling for admins with inadequate role after updates

**BE:**
1. Removed `UpdateUserPrivilegeById` endpoint and pushed authorisation check for role update to `UpdateUser` endpoint -> more concise `updateUserDto` and simpler front-end logic for updating users as admin
2. Added role guard for `/questions` endpoints to align authorisation (to revert/bypass if it helps with development)

### To-Do
- [x] Fix sorting of columns in table
- [x] Fix filtering of users in table

### CODE REVIEW

**Comment by javinchua** [comment]
would it be better to abstract the table into a separate component then users and questions can reuse it? cause i think in this implementation there would be quite abit of repeated code right

**Review by jonyeokj — CHANGES_REQUESTED** [review]
Add a check so that the admin cannot edit the information of the account that they are making the change from. A lot of buggy behaviour (table not updating, the whole page crashes, etc.) because of this

Also, this problem also happens when changing another admin's permission while that admin is viewing the /change-users page, causing him to crash when performing any actions. In fact, any other information changed (username, email) will cause said admin to crash.

Screenshot 2024-11-06 at 1 56 34 AM

However, I think this is a non-problem due to the unlikeliness of such an event, so preventing a user from editing himself is enough. If you want to prevent this though, you can prevent admins from editing other admins altogether.

File: project/packages/dtos/src/users.ts [hunk 1]
```diff
@@ -21,21 +23,28 @@ export const userCollectionSchema = z.object({
...
+export const sortUsersQuerySchema = z.object({
+  field: z.string(),
+  order: z.enum(["asc", "desc"]),
+});
+
 export const userFiltersSchema = z.object({
-  email: emailSchema,
-  username: z.string().optional(),
+  email: emailSchema.optional(),
+  username: usernameSchema.optional(),
```
Comment (jonyeokj) [inline project/packages/dtos/src/users.ts:33]: Using the usernameSchema here causes the filter in the table to make a false check on whether there are <4 characters when the user is filtering

**Review by jonyeokj — CHANGES_REQUESTED** [review]
Filtering doesn't work because of the schema

**Comment by gongg21** [comment]
@jonyeokj Fixes to bugs made:
- Active admin cannot view/update/delete themselves on /manage-users page anymore to deconflict
- Active admin can still view/update/delete other admins:
     - For other admins, they can see changes after refreshing only
     - If privilege changed to User, the ex-Admins will be redirected for any action taken on /manage-users page, if they were originally on it prior to the update (low chance scenario, addressed simply with redirection).

File: project/packages/dtos/src/users.ts (hunk 1 above)
Comment (gongg21) [reply project/packages/dtos/src/users.ts]: updated!

**Comment by jonyeokj** [comment]
Don't remove the view option of users on themselves on the table, it makes the row numbering in the table inconsistent - Just hide the options to update or delete.
Also, debounce the filter username option because the query happens immediately upon input change, making UX frustrating.

Other than that, LGTM

**Review by jonyeokj — APPROVED** [review]

**Pull request merged**